### PR TITLE
ARROW-3834: [Doc] Merge C++ and Python documentation

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
@wesm @pitrou Would love to get your feedback on this early as it involves some work. The C++ part works nice with the XML generated by `doxygen` but one needs to explicitly list the parts that should show up in Sphinx. There is also an autoindex variant for C++ but that produces a wall of errors. I'm quite happy with the approach.

@kou As you mentioned in the mail thread, GLib needs the XML for GObject introspection to work, so I'm not going to touch that. I would still add top-level sections for the languages based on it and link to the generated API docs. Would that be ok for you as a first shot?

**Comment by kou** [comment]
> Would that be ok for you as a first shot?

Yes!

**Comment by pitrou** [comment]
I almost never read the C++ docs (reading the .h files instead), so I'm not sure how people like to use it currently. I like this PR on the principle.

**Comment by wesm** [comment]
I will take a look when I can. We might go ahead and promote this to the top level of the repository to start having more documentation across languages

**Comment by kszucs** [comment]
I'm not sure about this. We should either move everything (including go, rust, java, ruby, etc.) to sphinx and make `doc` a top level directory or use the standard documentation tool for that particular language.

**Comment by wesm** [comment]
I think there's two kinds of documentation:

* Prose explanations of things
* Generated / browsable API documentation

In Python we are using Sphinx for both. Other projects only are doing the latter.

I would be in favor of having a top level doc/ directory where we develop prose documentation for the entire Arrow project including some of its programming language implementations. How we deal with generated API documentation may be on a case by case basis, but the doc index page should have links to every API listing that exists (Java, GLib, Python, Ruby, JavaScript, etc.). Getting every project's API docs into a single Sphinx project is probably too much to ask

**Comment by pitrou** [comment]
I wanted to give this a try but it seems doc building currently fails. The Python API docs are not generated anymore.

**Comment by xhochy** [comment]
@pitrou I will have a look later but one thing I noticed is that it always helped to have the Python package build and installed using both:

```
python setup.py build_ext --build-type=$ARROW_BUILD_TYPE --with-parquet --with-plasma --inplace
python setup.py build_ext --build-type=$ARROW_BUILD_TYPE --with-parquet --with-plasma install
```

**Comment by xhochy** [comment]
@pitrou rebased and also seeing a problem with the API docs.

**Comment by xhochy** [comment]
These were issues with `sphinx>=1.8`. I'll push a fix soon.

**Comment by xhochy** [comment]
@kszucs This now also contains a docker setup for building the Sphinx documentation.

@wesm @pitrou This is ready for review / merge. It contains a very basic set of C++ documentation. I would like to work on that in a separate PR and get the things here merged to prevent further merge conflicts.

**Comment by pitrou** [comment]
Apparently there's a failed Travis-CI entry.

File: .gitignore
Comment (pitrou) [inline .gitignore]: Use a wildcard for the version number here?

File: ci/conda_env_python.yml
Comment (pitrou) [inline ci/conda_env_python.yml]: What is recommonmark?

File: doc/source/conf.py
Comment (pitrou) [inline doc/source/conf.py]: i'm -1 on including MarkDown docs in the Sphinx source tree. We should convert those documents to reST format. It's a bit of work, but it will be much nicer afterwards.

**Review by pitrou — CHANGES_REQUESTED** [review]

File: ci/conda_env_python.yml
Comment (kszucs) [reply ci/conda_env_python.yml]: `conda_env_cpp.yml` and `conda_env_python.yml` files contain arrow-cpp's and pyarrow's dependencies and eventually will be used to spin up the development environment, see https://github.com/apache/arrow/pull/2836

I suggest to create a `ci/conda_env_sphinx.yml` or `conda_env_doc.yml` with the sphinx dependencies, including `doxygen` and `rsync` too.

File: cpp/Dockerfile
Comment (kszucs) [inline cpp/Dockerfile]: Move to `ci/conda_env_sphinx.yml`?

File: dev/docs/run_doc.sh
Comment (kszucs) [inline dev/docs/run_doc.sh]: Please rename this file according to the `Dockerfile` example above: `arrow/ci/docker_build_sphinx.sh` and call these scripts from the docker CMD instead.

File: doc/source/conf.py
Comment (kszucs) [inline doc/source/conf.py]: There is an unused `version` variable in the `conf.py` which We should set to `pyarrow.__version__` (suppose pyarrow is installed) and use here as `html_title = u'Apache Arrow {}'.format(version)`

File: docker-compose.yml
Comment (kszucs) [inline docker-compose.yml]:
```yml
    image: arrow:sphinx # or doc
    build:
      context: .
      dockerfile: doc/Dockerfile
```

**Review by kszucs — COMMENTED** [review]
@xhochy I can reorganize the docker related stuff too if You don't have time for it. 

Really minor: I'd use the plural `docs` instead of `doc` especially after merging multiple documentations under the same roof.

File: doc/source/conf.py
Comment (wesm) [reply doc/source/conf.py]: I don't think anyone has ever proposed converting the format docs to another markup language. In principle I don't see an issue with using reST because people will mostly be consuming the documents via the generated webpage. These aren't documents that are going to be frequently edited, anyway

Comment (xhochy) [reply doc/source/conf.py]: Ok, then I'll proceed and convert them to reST.

**Comment by xhochy** [comment]
@kszucs @pitrou Can you take a look at this again? I hope I have incorporated all comments.

**Comment by pitrou** [comment]
I'll take a look Monday.

File: python/README.md
Comment (kszucs) [inline python/README.md]: `../docs/requirements.txt`

File: docs/source/python/pandas.rst
Comment (kszucs) [inline docs/source/python/pandas.rst]: `Pandas`

File: dev/release/run-rat.sh
Comment (kszucs) [inline dev/release/run-rat.sh]: Don't hardcode rat's version number.

File: docs/requirements.txt
Comment (kszucs) [inline docs/requirements.txt]: I guess recommonmark is no longer needed.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
One thing: I think `Benchmarks.md` isn't part of the documentation tree at all. It should be in the top-level `python` directory (it's really some kind of README).

**Comment by xhochy** [comment]
Will merge on green build and follow-up with `Benchmarks.md` in a separate PR to unblock other branches that also touch the documentation: https://issues.apache.org/jira/browse/ARROW-3932

File: dev/release/run-rat.sh
Comment (fsaintjacques) [inline dev/release/run-rat.sh]: I always recommend using the form `: ${RAT_VERSION:=0.12}` as it allow you to test the script without changes, e.g. `$ RAT_VERSION=0.13 run-rat.sh` will test the script with `0.13 instead.

**Comment by wesm** [comment]
There was an anaconda.org timeout; I just restarted the failing build

**Review by wesm — APPROVED** [review]
+1. Thanks for doing this @xhochy! We may need to update various READMEs to point people to the format docs

**Comment by xhochy** [comment]
@wesm Fixed, so this is ready for re-review.

**Pull request closed without merging**
